File: c++/task3_philosophers.cpp
#include "task3_philosophers.h"
#include "benchmark_utils.h"
#include <iostream>
#include <thread>
#include <vector>
#include <mutex>
#include <random>
#include <chrono>
#include <algorithm>
#include <iomanip>
#include <condition_variable>

using namespace std::chrono_literals;

namespace task3 {

// Самодельный двоичный семафор для C++17
class BinarySemaphore {
private:
    std::mutex mtx;
    std::condition_variable cv;
    bool available;
    
public:
    BinarySemaphore(bool initial = true) : available(initial) {}
    
    void acquire() {
        std::unique_lock<std::mutex> lock(mtx);
        cv.wait(lock, [this]() { return available; });
        available = false;
    }
    
    void release() {
        {
            std::lock_guard<std::mutex> lock(mtx);
            available = true;
        }
        cv.notify_one();
    }
};

DiningPhilosophers::DiningPhilosophers(int num_philosophers, Strategy strategy)
    : num_philosophers_(num_philosophers), strategy_(strategy) {}

void DiningPhilosophers::philosopher_mutex(int id, int iterations, bool verbose) {
    static std::vector<std::mutex> forks(num_philosophers_);
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> think_dist(50, 200);
    std::uniform_int_distribution<> eat_dist(100, 300);
    
    int left_fork = id;
    int right_fork = (id + 1) % num_philosophers_;
    
    for (int i = 0; i < iterations; ++i) {
        // Размышление
        std::this_thread::sleep_for(std::chrono::milliseconds(think_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " размышляет (итерация " << i + 1 << ")\n";
        }
        
        // Захват вилок в определенном порядке для избежания deadlock
        if (id % 2 == 0) {
            forks[left_fork].lock();
            forks[right_fork].lock();
        } else {
            forks[right_fork].lock();
            forks[left_fork].lock();
        }
        
        // Еда
        std::this_thread::sleep_for(std::chrono::milliseconds(eat_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " ест спагетти (итерация " << i + 1 << ")\n";
        }
        
        // Освобождение вилок
        forks[left_fork].unlock();
        forks[right_fork].unlock();
    }
}

void DiningPhilosophers::philosopher_semaphore(int id, int iterations, bool verbose) {
    static std::vector<BinarySemaphore> forks(num_philosophers_);
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> think_dist(50, 200);
    std::uniform_int_distribution<> eat_dist(100, 300);
    
    int left_fork = id;
    int right_fork = (id + 1) % num_philosophers_;
    
    for (int i = 0; i < iterations; ++i) {
        // Размышление
        std::this_thread::sleep_for(std::chrono::milliseconds(think_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " размышляет (итерация " << i + 1 << ")\n";
        }
        
        // Захват вилок
        forks[left_fork].acquire();
        forks[right_fork].acquire();
        
        // Еда
        std::this_thread::sleep_for(std::chrono::milliseconds(eat_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " ест спагетти (итерация " << i + 1 << ")\n";
        }
        
        // Освобождение вилок
        forks[left_fork].release();
        forks[right_fork].release();
    }
}

void DiningPhilosophers::philosopher_try_lock(int id, int iterations, bool verbose) {
    static std::vector<std::mutex> forks(num_philosophers_);
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> think_dist(50, 200);
    std::uniform_int_distribution<> eat_dist(100, 300);
    std::uniform_int_distribution<> retry_dist(10, 50);
    
    int left_fork = id;
    int right_fork = (id + 1) % num_philosophers_;
    
    for (int i = 0; i < iterations; ++i) {
        // Размышление
        std::this_thread::sleep_for(std::chrono::milliseconds(think_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " размышляет (итерация " << i + 1 << ")\n";
        }
        
        // Попытка захвата вилок с повторными попытками
        bool has_forks = false;
        while (!has_forks) {
            if (forks[left_fork].try_lock()) {
                if (forks[right_fork].try_lock()) {
                    has_forks = true;
                } else {
                    forks[left_fork].unlock();
                    std::this_thread::sleep_for(std::chrono::milliseconds(retry_dist(gen)));
                }
            } else {
                std::this_thread::sleep_for(std::chrono::milliseconds(retry_dist(gen)));
            }
        }
        
        // Еда
        std::this_thread::sleep_for(std::chrono::milliseconds(eat_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " ест спагетти (итерация " << i + 1 << ")\n";
        }
        
        // Освобождение вилок
        forks[left_fork].unlock();
        forks[right_fork].unlock();
    }
}

void DiningPhilosophers::philosopher_arbitrator(int id, int iterations, bool verbose) {
    static std::mutex table_mutex;
    static std::vector<bool> forks_available(num_philosophers_, true);
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> think_dist(50, 200);
    std::uniform_int_distribution<> eat_dist(100, 300);
    
    int left_fork = id;
    int right_fork = (id + 1) % num_philosophers_;
    
    for (int i = 0; i < iterations; ++i) {
        // Размышление
        std::this_thread::sleep_for(std::chrono::milliseconds(think_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " размышляет (итерация " << i + 1 << ")\n";
        }
        
        // Запрос разрешения у арбитра (мьютекс стола)
        bool has_permission = false;
        while (!has_permission) {
            std::unique_lock<std::mutex> lock(table_mutex);
            
            if (forks_available[left_fork] && forks_available[right_fork]) {
                forks_available[left_fork] = false;
                forks_available[right_fork] = false;
                has_permission = true;
            }
            
            if (!has_permission) {
                lock.unlock();
                std::this_thread::sleep_for(10ms);
            }
        }
        
        // Еда
        std::this_thread::sleep_for(std::chrono::milliseconds(eat_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " ест спагетти (итерация " << i + 1 << ")\n";
        }
        
        // Возврат вилок
        {
            std::lock_guard<std::mutex> lock(table_mutex);
            forks_available[left_fork] = true;
            forks_available[right_fork] = true;
        }
    }
}

void DiningPhilosophers::philosopher_resource_hierarchy(int id, int iterations, bool verbose) {
    static std::vector<std::mutex> forks(num_philosophers_);
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> think_dist(50, 200);
    std::uniform_int_distribution<> eat_dist(100, 300);
    
    // Присваиваем порядковые номера вилкам
    int first_fork, second_fork;
    
    if (id == num_philosophers_ - 1) {
        // Последний философ берет вилки в обратном порядке
        first_fork = 0;
        second_fork = id;
    } else {
        first_fork = id;
        second_fork = (id + 1) % num_philosophers_;
    }
    
    // Убедимся, что first_fork имеет меньший номер
    if (first_fork > second_fork) {
        std::swap(first_fork, second_fork);
    }
    
    for (int i = 0; i < iterations; ++i) {
        // Размышление
        std::this_thread::sleep_for(std::chrono::milliseconds(think_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " размышляет (итерация " << i + 1 << ")\n";
        }
        
        // Захват вилок в порядке возрастания номеров
        forks[first_fork].lock();
        forks[second_fork].lock();
        
        // Еда
        std::this_thread::sleep_for(std::chrono::milliseconds(eat_dist(gen)));
        
        if (verbose && i < 10) {
            std::cout << "Философ " << id << " ест спагетти (итерация " << i + 1 << ")\n";
        }
        
        // Освобождение вилок
        forks[second_fork].unlock();
        forks[first_fork].unlock();
    }
}

void DiningPhilosophers::run_simulation(int iterations, bool verbose) {
    std::vector<std::thread> philosophers;
    
    std::string strategy_name;
    switch (strategy_) {
        case Strategy::MUTEX: strategy_name = "Мьютексы"; break;
        case Strategy::SEMAPHORE: strategy_name = "Семафоры"; break;
        case Strategy::TRY_LOCK: strategy_name = "Попытка захвата"; break;
        case Strategy::ARBITRATOR: strategy_name = "Арбитр"; break;
        case Strategy::RESOURCE_HIERARCHY: strategy_name = "Иерархия ресурсов"; break;
    }
    
    std::cout << "\n=== Задача обедающих философов ===\n";
    std::cout << "Философов: " << num_philosophers_ << "\n";
    std::cout << "Стратегия: " << strategy_name << "\n";
    std::cout << "Итераций: " << iterations << "\n";
    
    if (verbose && iterations > 10) {
        std::cout << "(Вывод ограничен первыми 10 итерациями)\n";
    }
    
    // Запуск философов
    for (int i = 0; i < num_philosophers_; ++i) {
        switch (strategy_) {
            case Strategy::MUTEX:
                philosophers.emplace_back(&DiningPhilosophers::philosopher_mutex, 
                                         this, i, iterations, verbose);
                break;
            case Strategy::SEMAPHORE:
                philosophers.emplace_back(&DiningPhilosophers::philosopher_semaphore, 
                                         this, i, iterations, verbose);
                break;
            case Strategy::TRY_LOCK:
                philosophers.emplace_back(&DiningPhilosophers::philosopher_try_lock, 
                                         this, i, iterations, verbose);
                break;
            case Strategy::ARBITRATOR:
                philosophers.emplace_back(&DiningPhilosophers::philosopher_arbitrator, 
                                         this, i, iterations, verbose);
                break;
            case Strategy::RESOURCE_HIERARCHY:
                philosophers.emplace_back(&DiningPhilosophers::philosopher_resource_hierarchy, 
                                         this, i, iterations, verbose);
                break;
        }
    }
    
    // Ожидание завершения
    for (auto& p : philosophers) {
        p.join();
    }
    
    std::cout << "\nСимуляция завершена успешно!\n";
}

void DiningPhilosophers::run_benchmark(int max_philosophers, int iterations) {
    std::cout << "\n=== Бенчмарк задачи обедающих философов ===\n";
    std::cout << "Тестируем разные стратегии и количество философов\n\n";
    
    std::vector<std::pair<std::string, double>> benchmark_results;
    
    std::vector<Strategy> strategies = {
        Strategy::MUTEX,
        Strategy::SEMAPHORE,
        Strategy::TRY_LOCK,
        Strategy::ARBITRATOR,
        Strategy::RESOURCE_HIERARCHY
    };
    
    std::vector<std::string> strategy_names = {
        "Мьютексы",
        "Семафоры",
        "Попытка захвата",
        "Арбитр",
        "Иерархия ресурсов"
    };
    
    std::vector<int> philosopher_counts = {5, 10, 20};
    
    for (int count : philosopher_counts) {
        if (count > max_philosophers) continue;
        
        for (size_t s = 0; s < strategies.size(); ++s) {
            std::string test_name = std::to_string(count) + "_философов_" + strategy_names[s];
            
            std::cout << "Тестируем: " << test_name << "... ";
            
            DiningPhilosophers dp(count, strategies[s]);
            
            Benchmark b(test_name, false);
            dp.run_simulation(iterations, false);
            
            double time = b.elapsed_microseconds();
            benchmark_results.emplace_back(test_name, time);
            
            std::cout << time << " мкс\n";
        }
    }
    
    Benchmark::save_to_csv(benchmark_results, "philosophers_benchmark.csv");
    std::cout << "\nБенчмарк завершен. Результаты сохранены в philosophers_benchmark.csv\n";
}

void run_philosophers() {
    std::cout << "\n=== Задание 3: Обедающие философы ===\n";
    std::cout << "Классическая задача синхронизации\n\n";
    
    int choice;
    std::cout << "Выберите режим:\n";
    std::cout << "1. Стандартная симуляция\n";
    std::cout << "2. Расширенный бенчмарк\n";
    std::cout << "Ваш выбор: ";
    std::cin >> choice;
    
    switch (choice) {
        case 1: {
            int num_philosophers, iterations, strategy_choice;
            
            std::cout << "\nВведите количество философов (2-20): ";
            std::cin >> num_philosophers;
            
            std::cout << "Введите количество итераций на философа (1-100): ";
            std::cin >> iterations;
            
            std::cout << "\nВыберите стратегию синхронизации:\n";
            std::cout << "1. Мьютексы (стандартная)\n";
            std::cout << "2. Семафоры\n";
            std::cout << "3. Попытка захвата (try_lock)\n";
            std::cout << "4. Арбитр (официант)\n";
            std::cout << "5. Иерархия ресурсов\n";
            std::cout << "Ваш выбор: ";
            std::cin >> strategy_choice;
            
            if (num_philosophers < 2) num_philosophers = 2;
            if (num_philosophers > 20) num_philosophers = 20;
            if (iterations < 1) iterations = 1;
            if (iterations > 100) iterations = 100;
            
            DiningPhilosophers::Strategy strategy;
            switch (strategy_choice) {
                case 1: strategy = DiningPhilosophers::Strategy::MUTEX; break;
                case 2: strategy = DiningPhilosophers::Strategy::SEMAPHORE; break;
                case 3: strategy = DiningPhilosophers::Strategy::TRY_LOCK; break;
                case 4: strategy = DiningPhilosophers::Strategy::ARBITRATOR; break;
                case 5: strategy = DiningPhilosophers::Strategy::RESOURCE_HIERARCHY; break;
                default: strategy = DiningPhilosophers::Strategy::MUTEX;
            }
            
            DiningPhilosophers dp(num_philosophers, strategy);
            
            Benchmark b("Симуляция обедающих философов");
            dp.run_simulation(iterations, true);
            break;
        }
        case 2: {
            int iterations;
            std::cout << "\nВведите количество итераций на философа (10-100): ";
            std::cin >> iterations;
            
            if (iterations < 10) iterations = 10;
            if (iterations > 100) iterations = 100;
            
            run_philosophers_benchmark();
            break;
        }
        default:
            std::cout << "Неверный выбор! Запускаю стандартную симуляцию...\n";
            DiningPhilosophers dp(5, DiningPhilosophers::Strategy::MUTEX);
            dp.run_simulation(10, true);
    }
}

void run_philosophers_benchmark() {
    std::cout << "\n=== Расширенный бенчмарк обедающих философов ===\n";
    
    int iterations;
    std::cout << "Введите количество итераций на философа (10-100): ";
    std::cin >> iterations;
    
    if (iterations < 10) iterations = 10;
    if (iterations > 100) iterations = 100;
    
    std::cout << "\nТестируем все стратегии...\n";
    
    DiningPhilosophers dp(5, DiningPhilosophers::Strategy::MUTEX);
    dp.run_benchmark(20, iterations);
}

} // namespace task3

File: go/task3_philosophers.go
package main

import (
	"context"
	"fmt"
	"math/rand"
	"sync"
	"time"

	"golang.org/x/sync/semaphore"
)

// Strategy определяет стратегию синхронизации
type Strategy int

const (
	MutexStrategy Strategy = iota
	SemaphoreStrategy
	TryLockStrategy
	ArbitratorStrategy
	ResourceHierarchyStrategy
)

// DiningPhilosophers реализует задачу обедающих философов
type DiningPhilosophers struct {
	numPhilosophers int
	strategy        Strategy
}

// NewDiningPhilosophers создает новую задачу философов
func NewDiningPhilosophers(numPhilosophers int, strategy Strategy) *DiningPhilosophers {
	return &DiningPhilosophers{
		numPhilosophers: numPhilosophers,
		strategy:        strategy,
	}
}

// philosopherMutex реализует философа с использованием мьютексов
func (dp *DiningPhilosophers) philosopherMutex(id, iterations int, verbose bool, forks []*sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()

	rng := rand.New(rand.NewSource(time.Now().UnixNano() + int64(id)))

	leftFork := id
	rightFork := (id + 1) % dp.numPhilosophers

	for i := 0; i < iterations; i++ {
		// Размышление
		thinkTime := 50 + rng.Intn(151) // 50-200 мс
		time.Sleep(time.Millisecond * time.Duration(thinkTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d размышляет (итерация %d)\n", id, i+1)
		}

		// Захват вилок в определенном порядке для избежания deadlock
		if id%2 == 0 {
			forks[leftFork].Lock()
			forks[rightFork].Lock()
		} else {
			forks[rightFork].Lock()
			forks[leftFork].Lock()
		}

		// Еда
		eatTime := 100 + rng.Intn(201) // 100-300 мс
		time.Sleep(time.Millisecond * time.Duration(eatTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d ест спагетти (итерация %d)\n", id, i+1)
		}

		// Освобождение вилок
		forks[leftFork].Unlock()
		forks[rightFork].Unlock()
	}
}

// philosopherSemaphore реализует философа с использованием семафоров
func (dp *DiningPhilosophers) philosopherSemaphore(id, iterations int, verbose bool, forks []*semaphore.Weighted, wg *sync.WaitGroup) {
	defer wg.Done()

	rng := rand.New(rand.NewSource(time.Now().UnixNano() + int64(id)))

	leftFork := id
	rightFork := (id + 1) % dp.numPhilosophers
	ctx := context.Background()

	for i := 0; i < iterations; i++ {
		// Размышление
		thinkTime := 50 + rng.Intn(151)
		time.Sleep(time.Millisecond * time.Duration(thinkTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d размышляет (итерация %d)\n", id, i+1)
		}

		// Захват вилок
		forks[leftFork].Acquire(ctx, 1)
		forks[rightFork].Acquire(ctx, 1)

		// Еда
		eatTime := 100 + rng.Intn(201)
		time.Sleep(time.Millisecond * time.Duration(eatTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d ест спагетти (итерация %d)\n", id, i+1)
		}

		// Освобождение вилок
		forks[leftFork].Release(1)
		forks[rightFork].Release(1)
	}
}

// philosopherTryLock реализует философа с попыткой захвата
func (dp *DiningPhilosophers) philosopherTryLock(id, iterations int, verbose bool, forks []*sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()

	rng := rand.New(rand.NewSource(time.Now().UnixNano() + int64(id)))

	leftFork := id
	rightFork := (id + 1) % dp.numPhilosophers

	for i := 0; i < iterations; i++ {
		// Размышление
		thinkTime := 50 + rng.Intn(151)
		time.Sleep(time.Millisecond * time.Duration(thinkTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d размышляет (итерация %d)\n", id, i+1)
		}

		// Попытка захвата вилок с повторными попытками
		hasForks := false
		for !hasForks {
			if forks[leftFork].TryLock() {
				if forks[rightFork].TryLock() {
					hasForks = true
				} else {
					forks[leftFork].Unlock()
					retryTime := 10 + rng.Intn(41) // 10-50 мс
					time.Sleep(time.Millisecond * time.Duration(retryTime))
				}
			} else {
				retryTime := 10 + rng.Intn(41)
				time.Sleep(time.Millisecond * time.Duration(retryTime))
			}
		}

		// Еда
		eatTime := 100 + rng.Intn(201)
		time.Sleep(time.Millisecond * time.Duration(eatTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d ест спагетти (итерация %d)\n", id, i+1)
		}

		// Освобождение вилок
		forks[leftFork].Unlock()
		forks[rightFork].Unlock()
	}
}

// philosopherArbitrator реализует философа с арбитром
func (dp *DiningPhilosophers) philosopherArbitrator(id, iterations int, verbose bool,
	tableMutex *sync.Mutex, forksAvailable []bool, wg *sync.WaitGroup) {
	defer wg.Done()

	rng := rand.New(rand.NewSource(time.Now().UnixNano() + int64(id)))

	leftFork := id
	rightFork := (id + 1) % dp.numPhilosophers

	for i := 0; i < iterations; i++ {
		// Размышление
		thinkTime := 50 + rng.Intn(151)
		time.Sleep(time.Millisecond * time.Duration(thinkTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d размышляет (итерация %d)\n", id, i+1)
		}

		// Запрос разрешения у арбитра
		hasPermission := false
		for !hasPermission {
			tableMutex.Lock()

			if forksAvailable[leftFork] && forksAvailable[rightFork] {
				forksAvailable[leftFork] = false
				forksAvailable[rightFork] = false
				hasPermission = true
			}

			if !hasPermission {
				tableMutex.Unlock()
				time.Sleep(10 * time.Millisecond)
			} else {
				tableMutex.Unlock()
			}
		}

		// Еда
		eatTime := 100 + rng.Intn(201)
		time.Sleep(time.Millisecond * time.Duration(eatTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d ест спагетти (итерация %d)\n", id, i+1)
		}

		// Возврат вилок
		tableMutex.Lock()
		forksAvailable[leftFork] = true
		forksAvailable[rightFork] = true
		tableMutex.Unlock()
	}
}

// philosopherResourceHierarchy реализует философа с иерархией ресурсов
func (dp *DiningPhilosophers) philosopherResourceHierarchy(id, iterations int, verbose bool, forks []*sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()

	rng := rand.New(rand.NewSource(time.Now().UnixNano() + int64(id)))

	// Присваиваем порядковые номера вилкам
	firstFork, secondFork := id, (id+1)%dp.numPhilosophers

	// Для последнего философа меняем порядок
	if id == dp.numPhilosophers-1 {
		firstFork, secondFork = 0, id
	}

	// Убедимся, что firstFork имеет меньший номер
	if firstFork > secondFork {
		firstFork, secondFork = secondFork, firstFork
	}

	for i := 0; i < iterations; i++ {
		// Размышление
		thinkTime := 50 + rng.Intn(151)
		time.Sleep(time.Millisecond * time.Duration(thinkTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d размышляет (итерация %d)\n", id, i+1)
		}

		// Захват вилок в порядке возрастания номеров
		forks[firstFork].Lock()
		forks[secondFork].Lock()

		// Еда
		eatTime := 100 + rng.Intn(201)
		time.Sleep(time.Millisecond * time.Duration(eatTime))

		if verbose && i < 10 {
			fmt.Printf("Философ %d ест спагетти (итерация %d)\n", id, i+1)
		}

		// Освобождение вилок
		forks[secondFork].Unlock()
		forks[firstFork].Unlock()
	}
}

// RunSimulation запускает симуляцию
func (dp *DiningPhilosophers) RunSimulation(iterations int, verbose bool) {
	var wg sync.WaitGroup

	strategyName := ""
	switch dp.strategy {
	case MutexStrategy:
		strategyName = "Мьютексы"
	case SemaphoreStrategy:
		strategyName = "Семафоры"
	case TryLockStrategy:
		strategyName = "Попытка захвата"
	case ArbitratorStrategy:
		strategyName = "Арбитр"
	case ResourceHierarchyStrategy:
		strategyName = "Иерархия ресурсов"
	}

	fmt.Println("\n=== Задача обедающих философов ===")
	fmt.Printf("Философов: %d\n", dp.numPhilosophers)
	fmt.Printf("Стратегия: %s\n", strategyName)
	fmt.Printf("Итераций: %d\n", iterations)

	if verbose && iterations > 10 {
		fmt.Println("(Вывод ограничен первыми 10 итерациями)")
	}

	// Инициализация ресурсов в зависимости от стратегии
	switch dp.strategy {
	case MutexStrategy:
		forks := make([]*sync.Mutex, dp.numPhilosophers)
		for i := 0; i < dp.numPhilosophers; i++ {
			forks[i] = &sync.Mutex{}
		}

		for i := 0; i < dp.numPhilosophers; i++ {
			wg.Add(1)
			go dp.philosopherMutex(i, iterations, verbose, forks, &wg)
		}

	case SemaphoreStrategy:
		forks := make([]*semaphore.Weighted, dp.numPhilosophers)
		for i := 0; i < dp.numPhilosophers; i++ {
			forks[i] = semaphore.NewWeighted(1)
		}

		for i := 0; i < dp.numPhilosophers; i++ {
			wg.Add(1)
			go dp.philosopherSemaphore(i, iterations, verbose, forks, &wg)
		}

	case TryLockStrategy:
		forks := make([]*sync.Mutex, dp.numPhilosophers)
		for i := 0; i < dp.numPhilosophers; i++ {
			forks[i] = &sync.Mutex{}
		}

		for i := 0; i < dp.numPhilosophers; i++ {
			wg.Add(1)
			go dp.philosopherTryLock(i, iterations, verbose, forks, &wg)
		}

	case ArbitratorStrategy:
		tableMutex := &sync.Mutex{}
		forksAvailable := make([]bool, dp.numPhilosophers)
		for i := 0; i < dp.numPhilosophers; i++ {
			forksAvailable[i] = true
		}

		for i := 0; i < dp.numPhilosophers; i++ {
			wg.Add(1)
			go dp.philosopherArbitrator(i, iterations, verbose, tableMutex, forksAvailable, &wg)
		}

	case ResourceHierarchyStrategy:
		forks := make([]*sync.Mutex, dp.numPhilosophers)
		for i := 0; i < dp.numPhilosophers; i++ {
			forks[i] = &sync.Mutex{}
		}

		for i := 0; i < dp.numPhilosophers; i++ {
			wg.Add(1)
			go dp.philosopherResourceHierarchy(i, iterations, verbose, forks, &wg)
		}
	}

	wg.Wait()
	fmt.Println("\nСимуляция завершена успешно!")
}

// RunBenchmark запускает бенчмарк
func (dp *DiningPhilosophers) RunBenchmark(maxPhilosophers, iterations int) {
	fmt.Println("\n=== Бенчмарк задачи обедающих философов ===")

	benchmarkResults := make([]BenchmarkResult, 0)

	strategies := []Strategy{MutexStrategy, SemaphoreStrategy, TryLockStrategy,
		ArbitratorStrategy, ResourceHierarchyStrategy}
	strategyNames := []string{"Мьютексы", "Семафоры", "Попытка захвата",
		"Арбитр", "Иерархия ресурсов"}

	philosopherCounts := []int{5, 10, 20}

	for _, count := range philosopherCounts {
		if count > maxPhilosophers {
			continue
		}

		for s, strategy := range strategies {
			testName := fmt.Sprintf("%d_философов_%s", count, strategyNames[s])

			fmt.Printf("Тестируем: %s... ", testName)

			dp := NewDiningPhilosophers(count, strategy)

			b := NewBenchmark(testName, false)
			dp.RunSimulation(iterations, false)

			time := b.StopSilent()
			benchmarkResults = append(benchmarkResults, BenchmarkResult{
				Name:             testName,
				TimeMicroseconds: time,
				TimeMilliseconds: time / 1000.0,
				TimeSeconds:      time / 1000000.0,
			})

			fmt.Printf("%.2f мкс\n", time)
		}
	}

	SaveToCSV(benchmarkResults, "philosophers_benchmark.csv")
	fmt.Println("\nБенчмарк завершен. Результаты сохранены в philosophers_benchmark.csv")
}

// RunPhilosophers запускает задание 3
func RunPhilosophers() {
	fmt.Println("\n=== Задание 3: Обедающие философы ===")
	fmt.Println("Классическая задача синхронизации\n")

	var choice string
	fmt.Println("Выберите режим:")
	fmt.Println("1. Стандартная симуляция")
	fmt.Println("2. Расширенный бенчмарк")
	fmt.Print("Ваш выбор: ")
	fmt.Scan(&choice)

	switch choice {
	case "1":
		var numPhilosophers, iterations, strategyChoice int

		fmt.Print("\nВведите количество философов (2-20): ")
		fmt.Scan(&numPhilosophers)

		fmt.Print("Введите количество итераций на философа (1-100): ")
		fmt.Scan(&iterations)

		fmt.Println("\nВыберите стратегию синхронизации:")
		fmt.Println("1. Мьютексы (стандартная)")
		fmt.Println("2. Семафоры")
		fmt.Println("3. Попытка захвата (try_lock)")
		fmt.Println("4. Арбитр (официант)")
		fmt.Println("5. Иерархия ресурсов")
		fmt.Print("Ваш выбор: ")
		fmt.Scan(&strategyChoice)

		if numPhilosophers < 2 {
			numPhilosophers = 2
		}
		if numPhilosophers > 20 {
			numPhilosophers = 20
		}
		if iterations < 1 {
			iterations = 1
		}
		if iterations > 100 {
			iterations = 100
		}

		var strategy Strategy
		switch strategyChoice {
		case 1:
			strategy = MutexStrategy
		case 2:
			strategy = SemaphoreStrategy
		case 3:
			strategy = TryLockStrategy
		case 4:
			strategy = ArbitratorStrategy
		case 5:
			strategy = ResourceHierarchyStrategy
		default:
			strategy = MutexStrategy
		}

		dp := NewDiningPhilosophers(numPhilosophers, strategy)

		b := NewBenchmark("Симуляция обедающих философов")
		dp.RunSimulation(iterations, true)
		b.Stop()

	case "2":
		var iterations int
		fmt.Print("\nВведите количество итераций на философа (10-100): ")
		fmt.Scan(&iterations)

		if iterations < 10 {
			iterations = 10
		}
		if iterations > 100 {
			iterations = 100
		}

		RunPhilosophersBenchmark()

	default:
		fmt.Println("Неверный выбор! Запускаю стандартную симуляцию...")
		dp := NewDiningPhilosophers(5, MutexStrategy)
		dp.RunSimulation(10, true)
	}
}

// RunPhilosophersBenchmark запускает расширенный бенчмарк
func RunPhilosophersBenchmark() {
	fmt.Println("\n=== Расширенный бенчмарк обедающих философов ===")

	var iterations int
	fmt.Print("Введите количество итераций на философа (10-100): ")
	fmt.Scan(&iterations)

	if iterations < 10 {
		iterations = 10
	}
	if iterations > 100 {
		iterations = 100
	}

	fmt.Println("\nТестируем все стратегии...")

	dp := NewDiningPhilosophers(5, MutexStrategy)
	dp.RunBenchmark(20, iterations)
}
